File: 프로그래머스/unrated/136797. 숫자 타자 대회/숫자 타자 대회.js
function solution(numbers) {
    const ws = Array.from(Array(10), () => Array(10).fill(1))
    for (let i = 0; i < 10; i++) {
        for (let j = 0; j<i; j++) {
            ws[j][i] = ws[i][j] = getWeigth(i,j)
        }
    }

    p = 4
    dp = [[6,0]]
    next = new Map()
    for (n of numbers) {
        next.clear()
        n = +n
        for (let [pp, v] of dp) {
            for (let [remain, replace] of [[p,pp], [pp,p]])
            if (remain !== n) {
                a = Math.min(next.get(remain)||Infinity, v+ws[replace][n])
                next.set(remain,a)
            }
        }
        dp = Array.from(next)
        p=n
    }
    return Array.from(next.values()).reduce((acc,cur) => Math.min(acc, cur), Infinity)
}

const getPos = (n) => [Math.floor((n-1)/3), (n-1)%3]

const getWeigth = (n1,n2) => {
    if (n1 === n2) return 1

    let  [[x1,y1],[x2,y2]] = [n1,n2].map(v => v===0?11:v).map(getPos)
    const [dx,dy] = [x1-x2,y1-y2].map(Math.abs)
    return Math.max(dx,dy)+dx+dy
}

File: 프로그래머스/unrated/136797. 숫자 타자 대회/숫자 타자 대회.py
def solution(numbers):
    key = {0:(3,1), **{n:divmod(n-1,3) for n in range(1,10)}}

    ws = [[1]*10 for _ in range(10)]
    for i in range(10):
        x_i, y_i = key[i]
        for j in range(i):
            x_j, y_j = key[j]
            dx , dy = map(abs, (x_i-x_j, y_i-y_j))
            ws[j][i] = ws[i][j] = max(dx,dy)+sum((dx,dy))
    dp = [(6,0)]
    p = 4
    for n in numbers:
        n = int(n)
        d = {}
        for pp,w in dp:
            if p != n:
                d[p] = min(d.get(p,10e9),w+ws[pp][n])
            if pp != n:
                d[pp] = min(d.get(pp,10e9),w+ws[p][n])
        p=n
        dp = [(k,v) for k,v in d.items()]

    return min((w for _,w in dp))

            